language/move-prover/bytecode/src/read_write_set_analysis.rs

```diff
@@ -0,0 +1,588 @@
+                    let addrs = state
+                        .locals
+                        .get_local(args[0])
+                        .expect("Unbound address local")
+                        .clone();
+                    let offset = Offset::global(mid, *sid, types.clone());
+                    let mut extended_aps: AbsAddr = AbsAddr::default();
+                    for p in addrs.iter() {
+                        match p {
+                            Addr::Footprint(ap) => {
+                                let mut extended_ap = ap.clone();
+                                extended_ap.add_offset(offset.clone());
+                                extended_aps.insert(Addr::Footprint(extended_ap.clone()));
+                                state.locals.update_access_path(extended_ap.clone(), None);
+                                state
+                                    .accesses
+                                    .update_access_path(extended_ap, Some(Access::Borrow))
+                            }
+                            Addr::Constant(c) => {
+                                let extended_ap = AccessPath::new_address_constant(
+                                    c.clone(),
+                                    mid,
+                                    *sid,
+                                    types.clone(),
+                                );
+                                extended_aps.insert(Addr::footprint(extended_ap));
+                            }
+                        }
+                    }
+                    state.locals.bind_local(rets[0], extended_aps)
+                }
+                MoveFrom(mid, sid, types) => {
+                    state.add_global_access(args[0], mid, *sid, types, Access::Write);
+                    state.remove_global(args[0], mid, *sid, types)
+                }
+                MoveTo(mid, sid, types) => {
+                    state.add_global_access(args[1], mid, *sid, types, Access::Write);
+                }
+                Exists(mid, sid, types) => {
+                    state.add_global_access(args[0], mid, *sid, types, Access::Read)
+                }
+                Function(mid, fid, types) => {
+                    let fun_id = mid.qualified(*fid);
+                    let global_env = self.cache.global_env();
+                    let callee_fun_env = global_env.get_function(fun_id);
+                    if let Some(callee_summary) = self.cache.get::<ReadWriteSetState>(fun_id) {
+                        state.apply_summary(callee_summary, args, types, rets)
+                    } else {
+                        // native fun. use handwritten model
+                        match (
+                            callee_fun_env.module_env.get_identifier().as_str(),
+                            callee_fun_env.get_identifier().as_str(),
+                        ) {
+                            ("BCS", "to_bytes") => {
```

Comment:
This should be factored out into its own function. Also we should likely add an enum somewhere centrally (TODO).

Reply:
Agree. Hopefully, most of these non-Novi specific native operations (e.g., vector stuff) will become bytecodes in time.

Reply:
Done